listing-rewards/program/src/execute_sale/mod.rs

```diff
@@ -0,0 +1,412 @@
+
+    /// CHECK: Not dangerous. Account seeds checked in constraint.
+    /// Free seller trade state PDA account encoding a free sell order.
+    #[account(
+        mut, 
+        seeds = [
+            PREFIX.as_bytes(),
+            seller.key().as_ref(),
+            auction_house.key().as_ref(), 
+            token_account.key().as_ref(),
+            auction_house.treasury_mint.as_ref(),
+            token_account.mint.as_ref(),
+            &0u64.to_le_bytes(), 
+            &execute_sale_params.token_size.to_le_bytes()
+        ], 
+        seeds::program = auction_house_program, 
+        bump = execute_sale_params.free_trade_state_bump
+    )]
+    pub free_seller_trade_state: UncheckedAccount<'info>,
+
+    /// CHECK: Verified through CPI
+    /// The auctioneer authority PDA running this auction.
+    #[account(
+        seeds = [
+            REWARD_CENTER.as_bytes(), 
+            auction_house.key().as_ref()
+        ], 
+        bump = reward_center.bump
+    )]
+    pub reward_center: Box<Account<'info, RewardCenter>>,
+
+    #[
+        account(
+            mut,
+            constraint = reward_center.token_mint == reward_center_reward_token_account.mint @ ListingRewardsError::MintMismatch
+        )
+    ]
+    /// The token account holding the reward token for the reward center.
+    pub reward_center_reward_token_account: Box<Account<'info, TokenAccount>>,
+
+    /// CHECK: Not dangerous. Account seeds checked in constraint.
+    /// The auctioneer PDA owned by Auction House storing scopes.
+    #[account(
+        seeds = [
+            AUCTIONEER.as_bytes(),
+            auction_house.key().as_ref(),
+            reward_center.key().as_ref()
+        ],
+        seeds::program = auction_house_program,
+        bump = auction_house.auctioneer_pda_bump
+    )]
+    pub ah_auctioneer_pda: Box<Account<'info, Auctioneer>>,
+
+    /// CHECK: Not dangerous. Account seeds checked in constraint.
+    #[account(
+        seeds = [
+            PREFIX.as_bytes(), 
+            SIGNER.as_bytes()
+        ], 
+        seeds::program = auction_house_program,
+        bump = execute_sale_params.program_as_signer_bump
+    )]
+    pub program_as_signer: UncheckedAccount<'info>,
+
+    /// Auction House Program
+    pub auction_house_program: Program<'info, AuctionHouseProgram>,
+    /// Token Program
+    pub token_program: Program<'info, Token>,
+    /// System Program
+    pub system_program: Program<'info, System>,
+    /// Associated Token Program
+    pub ata_program: Program<'info, AssociatedToken>,
+    /// Rent
+    pub rent: Sysvar<'info, Rent>,
+}
+
+pub fn handler(
+    ctx: Context<ExecuteSale>,
+    ExecuteSaleParams {
+        escrow_payment_bump,
+        free_trade_state_bump,
+        program_as_signer_bump,
+        price,
+        token_size,
+        ..
+    }: ExecuteSaleParams,
+) -> Result<()> {
+    let seller_listing = &mut ctx.accounts.listing;
+    let buyer_offer = &mut ctx.accounts.offer;
```

Comment:
With auctioneer its the responsibility of the third party program to enforce price matching which means we need an assertion that the offer amount is the same as the listing amount.

```rust
assert_eq!(offer.price, listing.price, ListingRewardsError::PriceMismatch)?;
```

Reply:
Ahh, good catch. I totally missed that.